ext/ASPExt/asp_uniform_tree_solver.jl

```diff
@@ -0,0 +1,116 @@
+        `statistics`: gives the statistics of the ASP solver
+    An ASPSolver is instantiated with a `grammar` and a `ruleNode`, automatically calling the `solve` function to retrieve all `solutions`.
+    The constraints of the `grammar` and the `ruleNode` are transformed to ASP rules. Then, Clingo_jll is used to generate all solutions for this Answer Set Program. These `solutions` can then be iterated. 
+    
+    To use the ASP Solver, Clingo_jll must be manually specified to be used, which automaticlaly loads the ASPExt extension module of HerbConstraints.
+    ```bash
+        julia> using Clingo_jll
+    ```
+"""
+mutable struct ASPSolver <: Solver
+    grammar::AbstractGrammar
+    ruleNode::Union{RuleNode,UniformHole,StateHole}
+    solutions::Vector{Dict{Int64,Int64}} #vector of dictionaries with key=node and value=matching rule index
+    isfeasible::Bool
+    statistics::Union{TimerOutput, Nothing}
+end
+
+"""
+    ASPSolver(grammar::AbstractGrammar, fixed_shaped_rulenode::AbstractRuleNode)
+"""
+function ASPSolver(grammar::AbstractGrammar, fixed_shaped_rulenode::AbstractRuleNode; with_statistics=false)
+    @assert !contains_nonuniform_hole(fixed_shaped_rulenode) "$(fixed_shaped_rulenode) contains non-uniform holes"
+    statistics = @match with_statistics begin
+        ::TimerOutput => with_statistics
+        ::Bool => with_statistics ? TimerOutput("ASP Solver") : nothing
+        ::Nothing => nothing
+    end
+    solver = ASPSolver(grammar, fixed_shaped_rulenode, Vector{Dict{Int32,Int32}}(), false, statistics)
+    solve(solver)
+    return solver
+end
+
+get_name(::ASPSolver) = "ASPSolver"
+
+"""
+    get_grammar(solver::ASPSolver)
+
+Get the grammar.
+"""
+function get_grammar(solver::ASPSolver)::AbstractGrammar
+    return solver.grammar
+end
+
+"""
+    get_rulenode(solver::ASPSolver)
+
+Get the RuleNode of the current program. This remains the same instance throughout the entire search.
+"""
+function get_rulenode(solver::ASPSolver)::AbstractRuleNode
+    return solver.ruleNode
+end
+
+"""
+    isfeasible(solver::ASPSolver)
+
+Returns true if no inconsistency has been detected.
+"""
+function isfeasible(solver::ASPSolver)
+    return solver.isfeasible
+end
+
+"""
+    solve(solver::ASPSolver)
+
+Generate all solutions for the given rulenode using ASP solver Clingo.
+"""
+function solve(solver::ASPSolver)
+    @timeit_debug solver.statistics "generate ASP RuleNode" begin end
+    string_rulenode, _ = rulenode_to_ASP(get_rulenode(solver), get_grammar(solver), 1)
+    constraints = grammar_to_ASP(get_grammar(solver))
+    asp_input = """
+%%% RuleNode
+$string_rulenode
+
+%%% Constraints
+$constraints
+
+%%% Query
+#show node/2.
+"""
+    buffer = IOBuffer(asp_input)
+    asp_output = IOBuffer()
+    @timeit_debug solver.statistics "run Clingo" begin end
+    run(pipeline(ignorestatus(`$(Clingo_jll.clingo()) --models 0`), stdin=buffer, stdout=asp_output))
```

Comment:
More of a question than anything, but does this mean that all of the solutions are extracted when an `ASPSolver` is constructed? If so, then what happens when we have a uniform tree with a huge number of solutions? I think this will take a long time to return the first tree.

I think that'd be fine for now, but if we want this to be a usable solver/iterator in the future, we'll need to query clingo for solutions one by one (on each call to `iterate(ASPIterator)`). This might mean we'd need to use `libclingo` directly rather than the command line interface. Maybe we can put this as a wishlist for a future hack-a-thon?